keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+
+Given the evacuation API is still a work in progress without any
+existing implementation, the annotation based eviction can be seen
+as an v1alpha1 implemenation of this proposal.
+Integration with the evacuation API as an v1alphaN or beta implementation.
+
+### Workflow Description
+
+This section outlines the workflow for enabling the proposed evictions
+in background that prevent workloads from being evicted right away
+without following additional eviction policies.
+
+#### Actors
+
+- Cluster Administrator: Responsible for configuring descheduling
+  policies and overseeing cluster operations.
+- User: Individuals or services running workloads within the cluster.
+- Descheduler: Component responsible for requesting pod evictions
+
+#### Workflow Steps
+
+1. Policy configuration
+   1. The cluster administrator configures the descheduler to enable the new functionality.
+2. A user deploys a workload through various nodes
+   1. A scheduler distributes the pods based on configured scheduling policies
+3. Eviction nomination
+   1. At the beginning of each descheduling cycle the descheduler increases
+      all the internal counters to take into account all pods that are subjects
+      to background eviction. Either by checking `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation,
+      list of evacuation requests or confronting the internal caches.
+   2. Each descheduling plugin nominates a set of pods to be evicted.
+   3. All the nominated pods are sorted based on the eviction-in-progress first priority.
+4. Pod eviction: the descheduler starts evicting nominated pods until a limit is reached
+   1. If an annotation based eviction is enabled:
+      1. If a pod is not annotated with `descheduler.alpha.kubernetes.io/request-evict-only` evict
+         the pod normally without additional handling.
+      2. If a pod is annotated with `descheduler.alpha.kubernetes.io/request-evict-only` evict the pod.
+         1. If the eviction fails intepret the error as a request for eviction in background.
+         2. Error code, resp. response text is checked to distinguish between
+            a genuine error and a confirmation of an eviction in background.
+   2. If the evacuation API is utilized create an evacution request
+
+### Notes/Constraints/Caveats (Optional)
+
+- The upstream eviction API does not currently support evictions in background
+  implemented by external components. For that, there's no community
+  dedicated error code nor a response text for this type of eviction.
+- The eviction in background error code will be temporary mapped to 429 (TooManyRequests).
```

Comment:
is it a good choice to use the error code that is already used by the eviction API?

Reply:
There's not many codes recognized by the eviction API. Plus, the expectation here is to use combination of both error code and response text. The response text makes the difference.

Reply:
If we were implemented this from ground up, having a different code would help to better distinguish the failures. But okay, I suppose it makes sense to use 429. An Eviction API sends this [when disruptions are not allowed](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L430). And also returns `RetryAfterSeconds`. Which is almost this case. And clients might handle it well.

Do we have any recommendations and use case for the `RetryAfterSeconds`?

Reply:
The current descheduler implementation relies on `--descheduling-interval`. There's no retry logic implemented. The next eviction/evacuation attempt is expected to be performed in the next descheduling cycle.